### PR TITLE
Add OpenSearch output connector

### PR DESCRIPTION
**Pull request by ppcad**
Add output connector for OpenSearch. See #161.

### CODE REVIEW

File: logprep/connector/opensearch/output.py [hunk 1]
```diff
@@ -0,0 +1,314 @@
...
+        currently_processed_cnt = self._processed_cnt + 1
+        if currently_processed_cnt == self._message_backlog_size:
+            try:
+                helpers.bulk(
+                    self._os,
+                    self._message_backlog,
+                    max_retries=self._max_retries,
+                    chunk_size=self._message_backlog_size,
+                )
+            except SerializationError as error:
```
Comment (urdit52) [inline logprep/connector/opensearch/output.py:162]: These Errors are all untested as of yet.

Comment (urdit52) [inline logprep/connector/opensearch/output.py:162]: These errors are untested.

File: logprep/connector/opensearch/output.py [hunk 2]
```diff
@@ -0,0 +1,314 @@
...
+                self._handle_serialization_error(error)
+            except ConnectionError as error:
+                self._handle_connection_error(error)
+            except BulkIndexError as error:
+                self._handle_bulk_index_error(error)
+            self._processed_cnt = 0
+
+            if self._input:
+                self._input.batch_finished_callback()
```
Comment (urdit52) [inline logprep/connector/opensearch/output.py:171]: this is untested.

File: logprep/connector/opensearch/output.py [hunk 3]
```diff
@@ -0,0 +1,314 @@
...
+        else:
+            self._processed_cnt = currently_processed_cnt
```
Comment (urdit52) [inline logprep/connector/opensearch/output.py:173]: this is untested.

File: logprep/connector/opensearch/output.py [hunk 4]
```diff
@@ -0,0 +1,314 @@
...
+        sent into an error index that should always accept all documents.
+        This can lead to a rebuild of the pipeline if this causes another exception.
+
+        Parameters
+        ----------
+        error : BulkIndexError
+           BulkIndexError to collect IndexErrors from.
+
+        """
+        error_documents = []
```
Comment (urdit52) [inline logprep/connector/opensearch/output.py:188]: 188-197 are untested.

File: logprep/connector/opensearch/output.py [hunk 5]
```diff
@@ -0,0 +1,314 @@
...
+        error : SerializationError
+           SerializationError for the error message.
+
+        Raises
+        ------
+        FatalOutputError
+            This causes a pipeline rebuild and gives an appropriate error log message.
+
+        """
+        raise FatalOutputError(f"{error.args[1]} in document {error.args[0]}")
```
Comment (urdit52) [inline logprep/connector/opensearch/output.py:237]: this is untested.

**Review by urdit52 — CHANGES_REQUESTED** [review]
Thank you very much for submitting this.

Could we please increase the test coverage back up to 100%?

Also, the code is almost exactly the same as the ES connector. Should we maybe refactor this to avoid duplicated code?

File: logprep/connector/opensearch/output.py (hunk 2 above)
Comment (ppcad) [reply logprep/connector/opensearch/output.py:171]: I've added a test.

File: logprep/connector/opensearch/output.py (hunk 3 above)
Comment (ppcad) [reply logprep/connector/opensearch/output.py:173]: I've added a test.

File: logprep/connector/opensearch/output.py (hunk 5 above)
Comment (ppcad) [reply logprep/connector/opensearch/output.py:237]: I've added a test.

File: logprep/connector/opensearch/output.py (hunk 4 above)
Comment (ppcad) [reply logprep/connector/opensearch/output.py:188]: I've added a test.

File: logprep/connector/opensearch/output.py (hunk 1 above)
Comment (ppcad) [reply logprep/connector/opensearch/output.py:162]: I've added a test.

File: logprep/connector/opensearch/output.py [hunk 6]
```diff
@@ -0,0 +1,314 @@
...
+        self._input = None
+
+        self._hosts = hosts
+        self._default_index = default_index
+        self._error_index = error_index
+        self._max_retries = max_retries
+
+        ssl_context = create_default_context(cafile=cert) if cert else None
+        if ssl_context and check_hostname is not None:
+            ssl_context.check_hostname = check_hostname
```
[the remark this answers is not in the document]
Comment (urdit52) [reply logprep/connector/opensearch/output.py:93]: I added a test.

**Review by urdit52 — APPROVED** [review]

**Pull request merged**